docker/oss-fuzz-builder/checkout_commit.py

```diff
@@ -0,0 +1,66 @@
+# Copyright 2020 Google LLC
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+# http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""Script for checking out a commit in an OSS-Fuzz project's repo."""
+import os
+import sys
+import subprocess
+
+
+def git(git_args, repo_dir):
+    """Execute a git command with |git_args| for the repo located in
+    |repo_dir|."""
+    command = ['git', '-C', repo_dir] + git_args
+    return subprocess.run(command, check=True)
+
+
+def fetch_unshallow(repo_dir):
+    """Gets the history of |repo_dir|."""
+    shallow_file = os.path.join(repo_dir, '.git', 'shallow')
+    if os.path.exists(shallow_file):
+        git(['fetch', '--unshallow'], repo_dir)
+
+
+def checkout_repo_commit(commit, repo_dir):
+    """Checkout |commit| in |repo_dir|."""
+    fetch_unshallow(repo_dir)
+    # TODO(metzman): Figure out if we need to run clean.
+    return git(['checkout', '-f', commit], repo_dir)
+
+
+def main():
+    """Check out an OSS-Fuzz project repo."""
+    commit = sys.argv[1]
+    if not commit:
+        print('Not checking out commit.')
+        return 0
+    # The SRC env var which is the /src directory in the oss-fuzz builder image
```

Comment:
Consider adding some documentation, e.g. to the top of this file, describing what this script does. It's still unclear to me because I don't know what is the input is. E.g., is `commit` a commit of github.com/google/oss-fuzz, or some upstream project repo? what is in the "/src directory in the oss-fuzz builder"? What are the directories under src_dir that we are iterating over, and why? What is the "correct project repo directory" and why do we need to search for it?

Reply:
`commit` is the commit for the project repo not the oss-fuzz repo. The search is over all the directories to try git checkout using the commit hash, so it's successful only in the project repo (the correct directory) to infer the path to the repo without the integrator having to specify. I'll add the same detail in the documentation.

Reply:
Thanks for the explanation! This feels pretty convoluted to me. I wonder why don't we checkout the commit right away after cloning the repo in the Dockerfile, ie
``` 
RUN git clone ... https://github.com/madler/zlib.git
RUN git checkout ...
```

Reply:
That's a good point. As far as I understand, this can be done without intervention by the integrator by modifying the dockerfile using code if the name of the project repo directory is same as the project. However, that would circle back to the issue because of which we check all directories to checkout the commit (because the name of the directory may not be same as the project). So to get the name from the dockerfile and appending the checkout command would also become convoluted and may not look clean at all.

Reply:
I think doing the checkout in the dockerfile is a good goal, but I think this is fine for now. Doing it in the dockerfile is harder to get right and would probably require parsing the `git clone ...` line to figure out which is the correct clone line to modify (many oss-fuzz dockerfiles clone multiple repos)?

Reply:
Okay. I'll add it to `checkout_commit.py`.